### PR TITLE
Deployment script module - AKS Run Command

### LINKED ISSUE #73: [Module Proposal]: Azure CLI DeploymentScript Module - AKS Run Command
### Have you checked this module does not already exist in the registry?

Yes

### Why is the module needed?

Extension of #60 and #69 

[DeploymentScripts](https://docs.microsoft.com/en-us/azure/templates/microsoft.resources/deploymentscripts?tabs=bicep) typically require several Azure resources to be deployed together to execute a meaningful deployment (ID, RBAC and the Script Itself). It lends itself to be in a registry because of this nature.

As well as serving as a meaningful module for customers, it will be directly leveraged in the [Azure/Aks-Construction repo](https://github.com/Azure/AKS-Construction) to reduce complexity in post-deployment automation steps;

- Using the AZ CLI to leverage the [ AKS Run Command](https://azure.microsoft.com/en-gb/updates/general-availability-aks-run-command/) to install CertManager on the cluster
- Using the AZ CLI to leverage the [ AKS Run Command](https://azure.microsoft.com/en-gb/updates/general-availability-aks-run-command/) to install External-DNS on the cluster
- Using the AZ CLI to leverage the [ AKS Run Command](https://azure.microsoft.com/en-gb/updates/general-availability-aks-run-command/) to install a sample application (the Azure Vote App) on the cluster
- Using the AZ CLI to leverage the [ AKS Run Command](https://azure.microsoft.com/en-gb/updates/general-availability-aks-run-command/) to query the Kubernetes events for `Warnings`

### Module path

deployment-scripts/aks-run-command

### Describe the module

A single bicep file that will deploy a managed identity, apply rbac permissions on existing resources and run an Azure CLI script.

The Azure CLI script will

- Invoke the AKS Run Command, relaying the command it has been provided with

### PR DESCRIPTION
**Pull request by Gordonby**
## Description

Adds a new module which leverages the AKS Run Command. The Run Command is particularly useful for interacting with private clusters.

Closes #73 

If you haven't already, read the full [contribution guide](../CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, edit the PR description and run through the relevant checklist below.

## Adding a new module

- [x] A proposal has been submitted and approved.
- [x] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [x] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

### CODE REVIEW

File: modules/deployment-scripts/aks-run-command/test/main.test.bicep [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  name: 'helmContour'
+  params: {
+    managedIdentityName: 'helmContourIngress'
+    rbacRolesNeeded:[
+      contributor
+      rbacWriter
+    ]
+    aksName: prereq.outputs.aksName
+    location: location
+    commands: [
```
Comment (alex-frankel) [inline modules/deployment-scripts/aks-run-command/test/main.test.bicep:90]: Could we add this one as example in the read me? Seems helpful to show people they can run helm or kubectl commands.

**Review by alex-frankel — COMMENTED** [review]
This also looks good. Can you confirm the idempotency of this one as well?

**Comment by Gordonby** [comment]
> This also looks good. Can you confirm the idempotency of this one as well?

It is idempotent with the sample scripts.
It's possible to write a command that wouldn't be though.

File: modules/deployment-scripts/aks-run-command/test/main.test.bicep (hunk 1 above)
Comment (Gordonby) [reply modules/deployment-scripts/aks-run-command/test/main.test.bicep:90]: Will do.

**Comment by Gordonby** [comment]
@alex-frankel Regarding idempotency.
Since this module is proxying commands, it will be impossible to guarantee for a given scenario if it is idempotent.
Those familiar with kubectl and helm should know the correct commands to run.
eg.
```bash
helm install something
```
will fail when you run it the second time. Thats why the module examples use

```bash
helm upgrade --install something
```

The other thing to highlight is the outputs. Given the same input, the output will be different because it includes datetimes. Eg.

```json
[
    {
        "DeploymentScriptIndex": 0,
        "DeploymentScriptName": "AKS-Run-crtestpwzzt42byuv4m-kubectlgetnodes-0",
        "DeploymentScriptOutputs": {
            "exitCode": 0,
            "finishedAt": "2022-04-19T20:36:40+00:00",
            "id": "357de78fc46e4299be36390b6d5ad936",
            "logs": "NAME STATUS ROLES AGE VERSION\naks-np01-48759702-vmss000000 Ready agent 3m41s v1.21.9\naks-np01-48759702-vmss000001 Ready agent 3m46s v1.21.9\n",
            "provisioningState": "Succeeded",
            "reason": null,
            "startedAt": "2022-04-19T20:36:39+00:00"
        },
        "DeploymentScriptStartTime": "2022-04-19T20:33:34.6224943Z",
        "DeploymentScriptEndTime": "2022-04-19T20:37:17.0280388Z"
    }
]
```

**Review by alex-frankel — COMMENTED** [review]
@Gordonby - can you confirm that any relevant style changes from the certificates module has been reflected here? Looks good otherwise.

**Comment by Gordonby** [comment]
> @Gordonby - can you confirm that any relevant style changes from the certificates module has been reflected here? Looks good otherwise.

Confirmed.

The last 3 commits align the style with Brians guidance from #69

**Review by alex-frankel — APPROVED** [review]

**Comment by Gordonby** [comment]
I've got a peach of a repo waiting to use this module @shenglol - eta on the merge?

https://github.com/Gordonby/aksc-petclinic

**Comment by shenglol** [comment]
Sorry didn't notice that Alex approved it. Merging it now.

**Pull request merged**